prov/efa/src/efa_msg.c

```diff
@@ -0,0 +1,326 @@
+
+	assert(msg->iov_count <= base_ep->info->rx_attr->iov_limit);
+
+	wr = &base_ep->efa_recv_wr_vec[wr_index].wr;
+	wr->wr_id = (uintptr_t)msg->context;
+	wr->num_sge = msg->iov_count;
+	wr->sg_list = base_ep->efa_recv_wr_vec[wr_index].sge;
+
+	for (i = 0; i < msg->iov_count; i++) {
+		addr = (uintptr_t)msg->msg_iov[i].iov_base;
+
+		/* Set RX buffer desc from SGE */
+		wr->sg_list[i].length = msg->msg_iov[i].iov_len;
+		assert(msg->desc[i]);
+		efa_mr = (struct efa_mr *)msg->desc[i];
+		wr->sg_list[i].lkey = efa_mr->ibv_mr->lkey;
+		wr->sg_list[i].addr = addr;
+	}
+
+	base_ep->efa_recv_wr_vec[wr_index].wr.next = NULL;
+	if (wr_index > 0)
+		base_ep->efa_recv_wr_vec[wr_index - 1].wr.next = &base_ep->efa_recv_wr_vec[wr_index].wr;
+
+	wr_index++;
+
+	if (flags & FI_MORE)
+		return 0;
+
+	err = ibv_post_recv(qp->ibv_qp, &base_ep->efa_recv_wr_vec[0].wr, &bad_wr);
+	if (OFI_UNLIKELY(err)) {
+		/* On failure, ibv_post_recv() return positive errno.
+		 * Meanwhile, this function return a negative errno.
+		 * So, we do the conversion here.
+		 */
+		err = (err == ENOMEM) ? -FI_EAGAIN : -err;
+	}
+
+	base_ep->efa_recv_wr_index = 0;
+
+	return err;
+}
+
+static ssize_t efa_ep_recvmsg(struct fid_ep *ep_fid, const struct fi_msg *msg, uint64_t flags)
+{
+	struct efa_base_ep *base_ep = container_of(ep_fid, struct efa_base_ep, util_ep.ep_fid);
+
+	return efa_post_recv(base_ep, msg, flags);
+}
+
+static ssize_t efa_ep_recv(struct fid_ep *ep_fid, void *buf, size_t len,
+			   void *desc, fi_addr_t src_addr, void *context)
+{
+	struct efa_base_ep *base_ep = container_of(ep_fid, struct efa_base_ep, util_ep.ep_fid);
+	struct iovec iov;
+	struct fi_msg msg;
+
+	EFA_SETUP_IOV(iov, buf, len);
+	EFA_SETUP_MSG(msg, &iov, &desc, 1, src_addr, context, 0);
+
+	return efa_post_recv(base_ep, &msg, 0);
+}
+
+static ssize_t efa_ep_recvv(struct fid_ep *ep_fid, const struct iovec *iov, void **desc,
+			    size_t count, fi_addr_t src_addr, void *context)
+{
+	struct efa_base_ep *base_ep = container_of(ep_fid, struct efa_base_ep, util_ep.ep_fid);
+	struct fi_msg msg;
+
+	EFA_SETUP_MSG(msg, iov, desc, count, src_addr, context, 0);
+
+	return efa_post_recv(base_ep, &msg, 0);
+}
+
+static inline ssize_t efa_post_send(struct efa_base_ep *base_ep, const struct fi_msg *msg, uint64_t flags)
+{
+	struct efa_qp *qp = base_ep->qp;
+	struct efa_conn *conn;
+	struct ibv_sge sg_list[2];  /* efa device support up to 2 iov */
+	struct ibv_data_buf inline_data_list[2];
+	size_t len, i;
+	int ret = 0;
+
+	dump_msg(msg, "send");
+
+	conn = efa_av_addr_to_conn(base_ep->av, msg->addr);
+	assert(conn && conn->ep_addr);
+
+	assert(msg->iov_count <= base_ep->info->tx_attr->iov_limit);
+
+	len = ofi_total_iov_len(msg->msg_iov, msg->iov_count);
+	assert(len <= base_ep->info->ep_attr->max_msg_size);
+
+	if (!base_ep->is_wr_started) {
+		ibv_wr_start(qp->ibv_qp_ex);
+		base_ep->is_wr_started = true;
+	}
+
+	qp->ibv_qp_ex->wr_id = (uintptr_t)msg->context;
+	if (flags & FI_REMOTE_CQ_DATA) {
```

Comment:
Since we are here, I think we should update this https://github.com/ofiwg/libfabric/blob/main/prov/efa/src/efa_prov_info.c#L78 to the efa device imm data size, which is 4

https://github.com/ofiwg/libfabric/commit/ead305192bd78d2395c64b99b02f79b2f16be8b2